web: adopt @goauthentik/theme tokens in the app styles (#22523)
goauthentik/authentik: website/docs/sys-mgmt/brands/design-language.mdx

---
title: Design Language
slug: /brands/design-language
sidebar_label: Design Language
---

authentik's design language is the small set of stable conventions — colors, spacing, typography, motion, shadows, stacking — that every screen in the interface uses to render itself. Authoring those decisions in one place means a brand operator can override a handful of CSS variables and get a coherent result across every flow, dialog, and admin view, without targeting individual components.

This page describes the **what** and **why** of the design language. For practical "how do I change X" recipes, see [Custom CSS](./custom-css.mdx).

:::warning Advanced topic

Custom CSS is an advanced feature. Improper use may lead to unexpected behavior or visual issues. Test changes in a safe environment before applying them to production.

:::

> ## How the design system is organised

Reviewer: ```suggestion
## How the design system is organized
```

| Layer                      | Example                                                       | What it controls                                                                | Recommended audience                      |
| -------------------------- | ------------------------------------------------------------- | ------------------------------------------------------------------------------- | ----------------------------------------- |
| **Semantic tokens**        | `--ak-color-primary`, `--ak-space-md`, `--ak-shadow-lg`       | The shared design decisions every component reads. Affects the whole interface. | Brand custom CSS, themes.                 |
| **Component-local tokens** | `--ak-c-login__main--BackgroundColor`, drawer width variables | One component's tunable knobs.                                                  | When a single component needs adjustment. |
| **CSS parts**              | `ak-flow-executor::part(locale-select)`                       | Structural hooks a component intentionally exposes.                             | Targeted layout or visibility changes.    |

> Most brand customisation stays at the first layer. The lower layers are escape hatches.

Reviewer: ```suggestion
Most brand customization stays at the first layer. The lower layers are escape hatches.
```

## Token groups

> The semantic surface is around 60 tokens, organised by concern. Every token name follows the pattern `--ak-{group}-{member}`.

Reviewer: ```suggestion
The semantic surface consists of around 60 tokens, organised by concern. Every token name follows the pattern `--ak-{group}-{member}`.
```

| `duration`     | `normal`                                                                                                                                                                                                          | The default transition duration. Adjust to make the whole interface feel snappier or calmer.                   |
| `easing`       | `standard`                                                                                                                                                                                                        | The default cubic-bezier curve.                                                                                |
| `z-index`      | `xs` through `2xl`                                                                                                                                                                                                | The stacking scale matches PatternFly's tiers (100..600).                                                      |

> 

Reviewer 2: Breakpoints?
Opacities?
Grids?
Any discussion of container queries, now that they're considered baseline widely available?
Gradients? 

... just some things to think about.

Author: Added a "What is deliberately not a token" section to `design-language.mdx` covering these: breakpoints (components use container queries — baseline now — and the remaining media queries follow PF grid tiers), opacities (baked into composed values; a standalone scale layers unpredictably against themed surfaces), grids/layout (component behavior, not brand identity), and gradients (the existing ones — skeleton shimmer, scroll shadows — are component implementation details composed from semantic colors). Container queries got their mention; nothing tokenized until usage demands it.

- **A single size scale: `xs`, `sm`, `md`, `lg`, `xl`, `2xl`, `3xl`, `4xl`.** Used consistently across `space`, `font.size`, `radius`, `border-width`, `shadow`, and `z-index`. There are no descriptive scales like `extraTight` or `loose`.
- **Spelled-out group names.** `z-index`, not `z`. `border-width`, not `bw`. Discoverable in IDE autocomplete and brand custom CSS without having to memorise abbreviations.
- **Private variables use a different prefix.** Variables starting with `--_` or `--ak-c-` are component-local and undocumented. They may change between authentik releases. Avoid targeting them in custom CSS unless a release explicitly documents one as stable.

## Light and dark themes

The design language ships with two themes out of the box. Both are toggled by the `data-theme` attribute on `html`:

```css title="Switching themes"
html[data-theme="dark"] {
    /* dark theme overrides apply here */
}
```

> A third theme — `reduced` — is reserved for users who explicitly opt into reduced motion (in addition to the browser-level `prefers-reduced-motion` media query, which fires automatically). authentik respects both, so most operators don't need to do anything to support reduced motion users.

Reviewer: ```suggestion
A third theme — `reduced` — is reserved for users who explicitly opt into reduced motion (in addition to the browser-level `prefers-reduced-motion` media query, which fires automatically). authentik respects both, so most operators do not need to do anything to support reduced motion users.
```

When you override semantic tokens in your custom CSS, follow the same pattern: provide light values under `:root`, then dark values under `html[data-theme="dark"]`. See [Custom CSS · Theme-based styling](./custom-css.mdx) for a worked example.

## Consuming tokens outside authentik

> If your organisation has its own design tooling, you can pull authentik's tokens in directly. The token system is published as the [`@goauthentik/theme`](https://www.npmjs.com/package/@goauthentik/theme) npm package, with three output formats:

Reviewer: ```suggestion
If your organization has its own design tooling, you can pull authentik's tokens in directly. The token system is published as the [`@goauthentik/theme`](https://www.npmjs.com/package/@goauthentik/theme) npm package, with three output formats:
```

| `@goauthentik/theme/index.css`                 | CSS           | Drop into a stylesheet to mirror authentik's tokens in another web surface. |
| `@goauthentik/theme/color.css` (and per group) | CSS           | Cherry-pick one category — for example, just colors, or just shadows.       |
| `@goauthentik/theme/tokens.dtcg.json`          | DTCG 2025.10  | Import into Tokens Studio, Terrazzo, Figma, or any other DTCG-aware tool.   |
| `@goauthentik/theme/tokens.dtcg.resolver.json` | DTCG resolver | Companion document that describes the `Dark` and `Reduced` theme modifiers. |

The DTCG JSON files are intended to be read or hand-edited; they're formatted and validated as part of the package's build.

## Stability

The `--ak-*` semantic tokens are the public API. authentik treats them as a stable contract within a major version:

- Names won't be renamed without a deprecation cycle.
- Values may shift to follow design improvements, but the _role_ a token plays (primary, surface, text-muted, etc.) is stable.
- New tokens may be added.
- Tokens may be removed only after a deprecation notice in the release notes.

> Anything _not_ in the `--ak-*` namespace — `--ak-c-*`, `--pf-*`, `--_*`, internal class names — is implementation detail. Custom CSS targeting those names may break between releases. Treat them as escape hatches for situations the semantic surface doesn't cover yet, and report any cases where you find yourself reaching for them so we can promote those to public tokens.

Reviewer: ```suggestion
Anything _not_ in the `--ak-*` namespace — `--ak-c-*`, `--pf-*`, `--_*`, internal class names — is implementation detail. Custom CSS targeting those names may break between releases. Treat them as escape hatches for situations the semantic surface does not cover yet, and report any cases where you find yourself reaching for them so we can promote those to public tokens.
```
